staging/src/k8s.io/client-go/tools/cache/listwatch.go

```diff
@@ -93,95 +102,96 @@ func NewFilteredListWatchFromClient(c Getter, resource string, namespace string,
 	return &ListWatch{ListFunc: listFunc, WatchFunc: watchFunc}
 }
 
-func timeoutFromListOptions(options metav1.ListOptions) time.Duration {
-	if options.TimeoutSeconds != nil {
-		return time.Duration(*options.TimeoutSeconds) * time.Second
-	}
-	return 0
+// applyUserListOptions applies only those options that are allowed to be set in List and Watch for an Informer
+func applyUserListOptions(options metav1.ListOptions, to *metav1.ListOptions) {
+	/*
+	 * Can't copy:
+	 * - ResourceVersion
+	 * - Timeout
+	 * as these are used by informers internally
+	 */
+
+	to.LabelSelector = options.LabelSelector
+	to.FieldSelector = options.FieldSelector
+	to.IncludeUninitialized = options.IncludeUninitialized
 }
 
-// List a set of apiserver resources
-func (lw *ListWatch) List(options metav1.ListOptions) (runtime.Object, error) {
-	if !lw.DisableChunking {
-		return pager.New(pager.SimplePageFunc(lw.ListFunc)).List(context.TODO(), options)
-	}
-	return lw.ListFunc(options)
-}
-
-// Watch a set of apiserver resources
-func (lw *ListWatch) Watch(options metav1.ListOptions) (watch.Interface, error) {
-	return lw.WatchFunc(options)
-}
+// NewListWatchFromMethods creates a ListWatch object from a client supporting List and Watch methods.
```

Comment:
is this change necessary? this doesn't seem like an improvement to me

Reply:
I've been on both sides of this fence, swearing at Golang, ended up embracing this path. The function is covered by tests so presumably the only time this fails and fails consistently (with panic) is when you give it object not fulfilling the interface. I understand your point, but I wouldn't overestimate the compiler safety. Aren't informers (and others) erasing the type to `interface{}` as well to go around the fact that Golang doesn't have templates. We get other benefits and this is a way of living with the downsides.

Reply:
If informers precedent wouldn't be enough, we are also loosing compiler safety (in different way) in almost identical function
https://github.com/kubernetes/kubernetes/blob/1ee2ac07c10dd8096d37ace3627bdffe45c2b867/staging/src/k8s.io/client-go/tools/cache/listwatch.go#L74-L94
by treating `resource` as string instead of using the typed client.

Reply:
I'm actually with @liggitt on this.
And I think there is difference between the example from above and what you're doing.

In the example above, we may potentially send a request to apiserver for non-existing resource, which will thrn result some error (which we need to handle anyway).

In the code below you're doing a lot of reflect.* operations which will be called at runtime (including panics).
I would really like to avoid getting that path. I would be very afraid of running such code as part of production.

Reply:
I'll drop it to move this PR forward

Reply:
Great - please ping me when this will be ready for next review pass.